### PR TITLE
Collect Azure SQL Server Cloud Databases

### PR DESCRIPTION
**Pull request by agrare**
This collects SQL Server Databases from Azure.  Azure has the concept of a "server" as well as a "database" in a more traditional server-based deployment compared to the more "cloud native" cloud database deployment.  Currently I am just looping through servers to fetch the databases, but we might want to add a `CloudDatabaseServer` model in the future to be able to model this.

https://github.com/ManageIQ/manageiq/issues/20221

### CODE REVIEW

File: app/models/manageiq/providers/azure/inventory/collector.rb [hunk 1]
```diff
@@ -295,6 +297,20 @@ def security_groups
     gather_data_for_this_region(@nsg)
   end
 
+  def sql_servers
+    @sql_servers ||= begin
+      @sqls.list_all.select do |server|
+        server.location == provider_region_description
```
Comment (agrare) [inline app/models/manageiq/providers/azure/inventory/collector.rb:303]: NOTE sql databases' locations are "US East 2" not "useast2"

File: app/models/manageiq/providers/azure/inventory/parser/cloud_manager.rb [hunk 2]
```diff
@@ -432,6 +433,13 @@ def image_operating_system(persister_miq_template, image)
     )
   end
 
+  def sql_databases
```
Comment (djberg96) [inline app/models/manageiq/providers/azure/inventory/parser/cloud_manager.rb:436]: Did you mean to keep byebug here?

Comment (agrare) [reply app/models/manageiq/providers/azure/inventory/parser/cloud_manager.rb:436]: No, still very WIP just wanted to throw this up

> Edit: Ah, I think it's WIP now. :)

Ah yeah I added the label but forgot to add it to the title

File: app/models/manageiq/providers/azure/inventory/collector/target_collection.rb [hunk 3]
```diff
@@ -251,6 +251,14 @@ def market_images
     end
   end
 
+  def sql_servers
+    []
+  end
+
+  def sql_databases
+    []
+  end
```
Comment (agrare) [inline app/models/manageiq/providers/azure/inventory/collector/target_collection.rb:260]: TODO will handle targeted refresh in a follow-up

**Comment by Fryguy** [comment]
@agrare A conflict occurred during the backport of this pull request to `morphy`.

If this pull request is based on another pull request that has not been marked for backport, add the appropriate labels to the other pull request. Otherwise, please create a new pull request direct to the `morphy` branch in order to resolve this.

Conflict details:

```diff
diff --cc app/models/manageiq/providers/azure/inventory/collector.rb
index d3d872a,6c4716b..0000000
--- a/app/models/manageiq/providers/azure/inventory/collector.rb
+++ b/app/models/manageiq/providers/azure/inventory/collector.rb
@@@ -217,6 -218,104 +219,107 @@@ class ManageIQ::Providers::Azure::Inven
      nil
    end
  
++<<<<<<< HEAD
++=======
+   def resource_groups
+     @resource_groups ||= collect_inventory(:resource_groups) { @rgs.list(:all => true) }
+   end
+ 
+   def flavors
+     collect_inventory(:series) do
+       begin
+         @vmm.series(@ems.provider_region)
+       rescue ::Azure::Armrest::BadGatewayException, ::Azure::Armrest::GatewayTimeoutException,
+              ::Azure::Armrest::BadRequestException => err
+         _log.error("Error Class=#{err.class.name}, Message=#{err.message}")
+         []
+       end
+     end
+   end
+ 
+   def availability_zones
+     collect_inventory(:availability_zones) { [::Azure::Armrest::BaseModel.new(:name => @ems.name, :id => 'default')] }
+   end
+ 
+   def instances
+     @instances_cache ||= collect_inventory(:instances) { gather_data_for_this_region(@vmm) }
+ 
+     instances_power_state_advanced_caching(@instances_cache) unless @instances_advanced_caching_done
+     @instances_advanced_caching_done = true
+ 
+     @instances_cache
+   end
+ 
+   # The underlying method that gathers these images is a bit brittle.
+   # Consequently, if it raises an error we just log it and move on so
+   # that it doesn't affect the rest of inventory collection.
+   #
+   def images
+     collect_inventory(:private_images) { gather_data_for_this_region(@sas, 'list_all_private_images') }
+   rescue ::Azure::Armrest::ApiException => err
+     _log.warn("Unable to collect Azure private images for: [#{@ems.name}] - [#{@ems.id}]: #{err.message}")
+     []
+   end
+ 
+   def managed_images
+     collect_inventory(:managed_images) { gather_data_for_this_region(@mis) }
+   end
+ 
+   # Collect marketplace image information if configured to do so. Normally
+   # users will specify images in their configuration file. If the option
+   # to collect marketplace images is selected, but there are no images
+   # specified in the configuration file, it will attempt to collect all
+   # marketplace images, which is an expensive operation.
+   #
+   def market_images
+     urns = options.market_image_urns
+ 
+     if urns
+       urns.collect do |urn|
+         publisher, offer, sku, version = urn.split(':')
+ 
+         ::Azure::Armrest::VirtualMachineImage.new(
+           :location  => manager.provider_region,
+           :publisher => publisher,
+           :offer     => offer,
+           :sku       => sku,
+           :version   => version,
+           :id        => urn
+         )
+       end
+     else
+       gather_data_for_this_region(@vmis)
+     end
+   end
+ 
+   def cloud_networks
+     gather_data_for_this_region(@vns)
+   end
+ 
+   def security_groups
+     gather_data_for_this_region(@nsg)
+   end
+ 
+   def sql_servers
+     @sql_servers ||= @sqls.list_all.select do |server|
+       # SqlServer instances have a "user friendly" location name
+       # e.g. "US East 2" rather than the more common "useast2" that the
+       # `gather_data_for_this_region` method is expecting
+       server.location == provider_region_description
+     end
+   end
+ 
+   def sql_databases
+     @sql_databases ||= sql_servers.flat_map do |sql_server|
+       @dbs.list_all(sql_server.name, sql_server.resource_group).map { |db| [sql_server, db] }
+     end
+   end
+ 
+   def load_balancers
+     @load_balancers ||= gather_data_for_this_region(@lbs)
+   end
+ 
++>>>>>>> cf6c300... Merge pull request #464 from agrare/collect_sql_databases
    protected
  
    attr_reader :record_limit, :enabled_deployments_caching
diff --cc app/models/manageiq/providers/azure/inventory/persister.rb
index 67f73dd,3444c20..0000000
--- a/app/models/manageiq/providers/azure/inventory/persister.rb
+++ b/app/models/manageiq/providers/azure/inventory/persister.rb
@@@ -25,6 -24,113 +25,116 @@@ class ManageIQ::Providers::Azure::Inven
      initialize_inventory_collections
    end
  
++<<<<<<< HEAD
++=======
+   def initialize_inventory_collections
+     initialize_tag_mapper
+     initialize_cloud_inventory_collections
+     initialize_network_inventory_collections
+   end
+ 
+   def initialize_cloud_inventory_collections
+     %i(availability_zones
+        cloud_databases
+        disks
+        flavors
+        hardwares
+        networks
+        operating_systems
+        resource_groups
+        miq_templates
+        vm_and_template_labels
+        vm_and_template_taggings
+        vms).each do |name|
+ 
+       add_cloud_collection(name)
+     end
+ 
+     add_auth_key_pairs
+ 
+     add_orchestration_stacks
+ 
+     # slightly different from amazon
+     %i(orchestration_stacks_resources
+        orchestration_stacks_outputs
+        orchestration_stacks_parameters
+        orchestration_templates).each do |name|
+ 
+       extra_properties = {:manager_ref => %i(stack ems_ref)} unless name == :orchestration_templates
+       add_cloud_collection(name, extra_properties)
+     end
+ 
+     # Custom processing of Ancestry
+     %i(vm_and_miq_template_ancestry
+        orchestration_stack_ancestry).each do |name|
+ 
+       add_cloud_collection(name)
+     end
+   end
+ 
+   def initialize_network_inventory_collections
+     %i(cloud_networks
+        cloud_subnet_network_ports
+        floating_ips
+        network_ports
+        network_routers
+        load_balancers
+        load_balancer_pools
+        load_balancer_pool_members
+        load_balancer_pool_member_pools
+        load_balancer_listeners
+        load_balancer_listener_pools
+        load_balancer_health_checks
+        load_balancer_health_check_members
+        security_groups).each do |name|
+ 
+       add_network_collection(name)
+     end
+ 
+     add_cloud_subnets
+ 
+     add_firewall_rules
+   end
+ 
+   # ------ IC provider specific definitions -------------------------
+ 
+   def add_auth_key_pairs(extra_properties = {})
+     add_cloud_collection(:auth_key_pairs, extra_properties) do |builder|
+       builder.add_properties(:model_class => ::ManageIQ::Providers::CloudManager::AuthKeyPair)
+       builder.add_properties(:manager_uuids => name_references(:key_pairs)) if targeted?
+     end
+   end
+ 
+   def add_orchestration_stacks
+     add_cloud_collection(:orchestration_stacks) do |builder|
+       builder.add_properties(:saver_strategy => 'default') # TODO(lsmola) can't batch unless we do smart batching
+     end
+   end
+ 
+   # TODO: Builder params
+   def add_cloud_subnets
+     add_network_collection(:cloud_subnets, :default_values => nil) do |builder|
+       builder.add_properties(:parent_inventory_collections => %i(cloud_networks))
+ 
+       builder.add_targeted_arel(
+         lambda do |inventory_collection|
+           manager_uuids = inventory_collection.parent_inventory_collections.flat_map { |c| c.manager_uuids.to_a }
+           inventory_collection.parent.cloud_subnets.joins(:cloud_network).where(
+             :cloud_networks => {:ems_ref => manager_uuids}
+           )
+         end
+       )
+     end
+   end
+ 
+   # TODO: Same as amazon?
+   def add_firewall_rules(extra_properties = {})
+     add_network_collection(:firewall_rules, extra_properties) do |builder|
+       builder.add_properties(:manager_ref_allowed_nil => %i(source_security_group))
+     end
+   end
+ 
++>>>>>>> cf6c300... Merge pull request #464 from agrare/collect_sql_databases
    private
  
    def case_sensitive_labels?
```

**Comment by agrare** [comment]
@Fryguy this backports cleanly with https://github.com/ManageIQ/manageiq-providers-azure/pull/454 and https://github.com/ManageIQ/manageiq-providers-azure/pull/444 applied

**Comment by Fryguy** [comment]
Backported to `morphy` in commit fc37882ea5364d2e2ade70fbfdb3a96c24319ad2.

```text
commit fc37882ea5364d2e2ade70fbfdb3a96c24319ad2
Author: Jason Frey <fryguy9@gmail.com>
Date:   Wed Aug 11 11:10:51 2021 -0400

    Merge pull request #464 from agrare/collect_sql_databases
    
    Collect Azure SQL Server Cloud Databases
    
    (cherry picked from commit cf6c300560ff7047bcb0b81a5db57a5883ff75d6)
```

**Pull request merged**